File: 10142 - Australian Voting/Australian Voting.py
import sys

lines = iter(sys.stdin.read().splitlines())

cases_n = int(next(lines))

for _ in range(cases_n):
    for line in lines:
        if line.isdigit():
            candidates_n = int(line)
            break
    
    candidates_list = [next(lines) for _ in range(candidates_n)]
    
    votes = []
    for i, line in enumerate(lines):
        vote = line.split()
        if not vote:
            if not line:
                break
        elif i < 1000:
            votes.append(list(map(lambda x: int(x)-1 , vote)))
    
    winners = []
    pos_president = [True]*candidates_n
    
    while not winners:
        counter = [0]*candidates_n
        for vote in votes:
            counter[vote[0]] += 1
    
        max_value = max(counter)
        if max_value > len(votes)/2.0:
            max_index = counter.index(max_value)
            winners.append(candidates_list[max_index])

        min_value = min(counter)
        if max_value == min_value:
            for i, posible in enumerate(pos_president):
                if posible:
                    winners.append(candidates_list[i])
        else:
            for i, count in enumerate(counter):
                if count == min_value:
                    pos_president[i] = False
                    for vote in votes:
                        vote.remove(i)
    
    for winner in winners:                 
        print(winner)

File: 10142 - Australian Voting/Australian Voting.java
import java.util.*;

class Main
{
    static Scanner input = new Scanner(System.in); 

    public static void main(String[] args)
    {
        int candidates_n;
        String candidates_list[];

        int votes_n, order_n;
        int votes[][] = new int [1000][20];
        boolean pos_president[];
        boolean winners[];
        int counter[];
        int pointers[];
        int maxim = 0, minim;

        int cases_n = input.nextInt(); input.nextLine();
        String line;
        for(int c=0; c < cases_n; c++)
        {
            do
                line = input.nextLine().trim();
            while(input.hasNext() && line.length() == 0);

            candidates_n = Integer.parseInt(line);
            candidates_list = new String [candidates_n];
            for(int i=0; i < candidates_n; i++)
                candidates_list[i] = input.nextLine().trim();
            votes_n = 0;
            do{
                line = input.nextLine().trim();
                if(line.length() == 0)
                    break;
                
                String[] integerStrings = line.split(" ");
                order_n = 0;
                for(String gap : integerStrings)
                    if(gap.length() > 0)
                    {
                        votes[votes_n][order_n] = Integer.parseInt(gap) - 1;
                        order_n++;
                    }
                votes_n++;
            }while(input.hasNext());

            pos_president = new boolean[candidates_n]; Arrays.fill(pos_president, true);
            winners = new boolean[candidates_n];
            pointers = new int [votes_n];
            
            while(true)
            {
                counter = new int[candidates_n];
                for(int i=0; i < votes_n; i++)
                {   
                    while(!pos_president[votes[i][pointers[i]]])
                        pointers[i]++;
                    maxim = ++counter[votes[i][pointers[i]]];
                    if(maxim <<1 > votes_n)
                    {   
                        winners[votes[i][pointers[i]]] = true;
                        break;
                    }
                }
                
                minim = maxim;
                for(int i=0; i < candidates_n; i++)
                    if(pos_president[i])
                    {
                        maxim = Math.max(maxim, counter[i]);
                        minim = Math.min(minim, counter[i]);
                    }
                if(minim == maxim)
                {   
                    winners = pos_president;
                    break;
                }
                else
                {
                    for(int i=0; i < candidates_n; i++)
                        if(pos_president[i] && minim == counter[i])
                            pos_president[i] = false;
                }

            }

            for(int i=0; i < candidates_n; i++)
                if (winners[i])
                    System.out.println(candidates_list[i]);
            if (c < cases_n-1)
                System.out.println();
        }
    }
}
